arch/eventloop_posix_tcp.c

```diff
@@ -490,6 +490,66 @@ TCP_registerListenSocketDomainName(UA_ConnectionManager *cm, const char *hostnam
     return UA_STATUSCODE_GOOD;
 }
 
+#ifdef _WIN32
+static
+UA_RegisteredFD *
+findRegisteredFD(TCPConnectionManager *cm, uintptr_t connectionId) {
+    UA_RegisteredFD *rfd, *rfd_tmp;
+    LIST_FOREACH_SAFE(rfd, &cm->fds, es_pointers, rfd_tmp) {
+
+        if (rfd->fd == connectionId) {
+            return rfd;
+        }
+    }
+    return NULL;
+}
+
+static
+UA_DelayedCallback *createExplicitlyDelayedShutdownCallback(UA_ConnectionManager *cm, UA_RegisteredFD *rfd, UA_Callback callback) {
+    UA_DelayedCallback *explicitlyDelayedShutdownCallback = (UA_DelayedCallback*) UA_malloc(sizeof(UA_DelayedCallback));
+    if (explicitlyDelayedShutdownCallback == NULL) {
+        return NULL;
+    }
+    explicitlyDelayedShutdownCallback->callback = callback;
+    explicitlyDelayedShutdownCallback->application = cm;
+    explicitlyDelayedShutdownCallback->data = rfd;
+    return explicitlyDelayedShutdownCallback;
+}
+
+static
+void delayedShutdownCallback(void* application, void* data) {
+
+    UA_ConnectionManager *cm = (UA_ConnectionManager*) application;
+    TCPConnectionManager *tcm = (TCPConnectionManager*)cm;
+    UA_RegisteredFD* rfd = (UA_RegisteredFD *) data;
+
+    UA_ByteString response = tcm->rxBuffer;
+    int ret = UA_recv(rfd->fd, (char *)response.data, response.length, MSG_DONTWAIT | MSG_PEEK);
+    if(ret == 0 || ret == -1) {
+        cm->connectionCallback(cm, (uintptr_t)rfd->fd, &rfd->context,
+                               UA_STATUSCODE_BADCONNECTIONCLOSED,
+                               0, NULL, UA_BYTESTRING_NULL);
+        TCP_close(tcm, rfd);
+    } else {
+
+        /* add the delayed shutdown callback again because the peeked data will be processed in the next el cycle */
+        UA_DelayedCallback *explicitlyDelayedShutdownCallback = createExplicitlyDelayedShutdownCallback(cm, rfd, delayedShutdownCallback);
```

Comment:
Move logging and `addDelayedCallback` into `createExplicitlyDelayedShutdownCallback`.
Currently you have duplicate code after both usage points of `createExplicitlyDelayedShutdownCallback`.
At this time, you don't need the callback argument.
Because you use the same callback in both cases.

Reply:
It was meant to be like this.  In `delayedShutdownCallback` the logging needs to be inline because I cannot return an error-code. Whereas in the other place I can return `UA_STATUSCODE_BADOUTOFMEMORY` and propagate the error and handle logging somewhere else.

As for the two lines of code that get the eventloop and add the callback, I would again say "explicit is better than implicit". The callback is created via `create...` and then added to the eventloop. If possible I like functions that serve as little purposes as possible.

Reply:
The calling function cannot handle an out-of-memory any better.
So we should handle it internally and not propagate the error outside of the current scope.

Normally we would reserve the memory already when the connection is created, so that the cleanup always works.
This pattern is also used in other places in the SDK already.

However, here the worst that can happen is a 120 second wait for the server shutdown.
And only on win32 where we are typically not memory constrained.
Hence, log the error with a warning and do not return an error code.

And with this, the two code paths can be merged as in the original review comment.